File: src/core/user/user.service.ts
import {
  BadRequestException,
  Injectable,
  NotFoundException,
  UnauthorizedException,
} from '@nestjs/common';
import { AuthService } from '../../shared/modules/auth/auth.service';
import { AuthResponseDto } from './dto/auth-response.dto';
import { CreateUserDto } from './dto/create-user.dto';
import { LoginUserDto } from './dto/login-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { User } from './entities/user.entity';
import { IUserRepository } from './repositories/user.repository';
import { DatabaseTransaction } from '../../shared/modules/database/database.transaction';

@Injectable()
export class UserService {
  public constructor(
    private readonly userRepository: IUserRepository,
    private readonly authService: AuthService,
  ) {}

  public async create(createUserDto: CreateUserDto): Promise<AuthResponseDto> {
    await this.validateEmail(createUserDto.email);

    const user = await this.userRepository.create(
      await this.createCreateUserData(createUserDto),
    );
    const token = await this.authService.generateToken({ id: user.id });

    delete user.password;
    return { token, user };
  }

  public async login(loginUserDto: LoginUserDto): Promise<AuthResponseDto> {
    const user = await this.userRepository.findByEmail(loginUserDto.email);
    if (!user) throw new UnauthorizedException('E-mail e/ou senha inválidos');

    const passwordMatch = await this.authService.comparePassword(
      loginUserDto.password,
      user.password!,
    );

    if (!passwordMatch)
      throw new UnauthorizedException('E-mail e/ou senha inválidos');

    const token = await this.authService.generateToken({ id: user.id });

    delete user.password;
    return { token, user };
  }

  public async getNewSession(id: string): Promise<AuthResponseDto> {
    const user = await this.findById(id);
    const newToken = await this.authService.generateToken({ id: user.id });
    delete user.password;
    return { user, token: newToken };
  }

  public async findById(
    id: string,
    transaction?: DatabaseTransaction,
  ): Promise<User> {
    const user = await this.userRepository.findById(id, transaction);
    if (!user) throw new NotFoundException('Conta não encontrada');
    return user;
  }

  public async findByEmail(email: string): Promise<User> {
    const user = await this.userRepository.findByEmail(email);
    if (!user) throw new NotFoundException('Conta não encontrada');
    delete user.password;
    return user;
  }

  public async update(id: string, updateUserDto: UpdateUserDto): Promise<void> {
    const user = await this.findById(id);

    if (updateUserDto.email && updateUserDto.email != user.email)
      await this.validateEmail(updateUserDto.email);

    await this.userRepository.update(
      id,
      await this.createUpdateUserData(updateUserDto),
    );
  }

  public async delete(id: string): Promise<void> {
    await this.findById(id);
    await this.userRepository.delete(id);
  }

  private async validateEmail(email: string): Promise<void> {
    const user = await this.userRepository.findByEmail(email);
    if (user)
      throw new BadRequestException('Esse e-mail já está sendo utilizado');
  }

  private async createCreateUserData(
    createUserDto: CreateUserDto,
  ): Promise<CreateUserDto> {
    return {
      email: createUserDto.email,
      name: createUserDto.name,
      password: await this.authService.hashPassword(createUserDto.password),
    };
  }

  private async createUpdateUserData(
    updateUserDto: UpdateUserDto,
  ): Promise<UpdateUserDto> {
    return {
      email: updateUserDto.email,
      name: updateUserDto.name,
      password: updateUserDto.password
        ? await this.authService.hashPassword(updateUserDto.password)
        : undefined,
    };
  }
}

File: src/main/java/dev/joaopdias/auronix/core/user/UserService.java
package dev.joaopdias.auronix.core.user;

import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import dev.joaopdias.auronix.core.user.dto.AuthResponseDto;
import dev.joaopdias.auronix.core.user.dto.CreateUserDto;
import dev.joaopdias.auronix.core.user.dto.LoginUserDto;
import dev.joaopdias.auronix.core.user.dto.UpdateUserDto;
import dev.joaopdias.auronix.core.user.entities.User;
import dev.joaopdias.auronix.shared.services.SecurityService;

@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private SecurityService securityService;

    public AuthResponseDto create(CreateUserDto createUserDto) {
        validateEmail(createUserDto.email());

        User user = new User();

        user.setEmail(createUserDto.email());
        user.setName(createUserDto.name());
        user.setPassword(securityService.hashPassword(createUserDto.password()));
        
        user = userRepository.save(user);
        
        String token = securityService.createJwt(user.getId());

        return new AuthResponseDto(token, user.toResponseDto());
    }

    public AuthResponseDto login(LoginUserDto loginUserDto) {
        User user = userRepository.findByEmail(loginUserDto.email());
        if (user == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email ou senha inválidos");

        boolean passwordMatches = securityService.matchesPassword(loginUserDto.password(), user.getPassword());
        if (!passwordMatches) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email ou senha inválidos");

        String token = securityService.createJwt(user.getId());

        return new AuthResponseDto(token, user.toResponseDto());
    }

    public AuthResponseDto decodeToken(String token) {
        UUID id = securityService.decodeJwt(token);
        User user = findById(id);
        String newToken = securityService.createJwt(user.getId());
        return new AuthResponseDto(newToken, user.toResponseDto());
    }

    public User findById(UUID id) {
        return userRepository.findById(id)
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Conta não encontrada"));
    }

    public void update(UUID id, UpdateUserDto updateUserDto) {
        User user = findById(id);

        if (updateUserDto.email() != null && !updateUserDto.email().equals(user.getEmail())) {
            validateEmail(updateUserDto.email());
            user.setEmail(updateUserDto.email());
        }

        if (updateUserDto.name() != null) 
            user.setName(updateUserDto.name());
        

        if (updateUserDto.password() != null) 
            user.setPassword(securityService.hashPassword(updateUserDto.password()));

        userRepository.save(user);
    }

    public void delete(UUID id) {
        User user = findById(id);
        userRepository.delete(user);
    }

    private void validateEmail(String email) {
        User user = userRepository.findByEmail(email);
        if (user != null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Esse email já está em uso");
    }
}
